src/include/daos_srv/daos_server.h

```diff
@@ -158,6 +142,87 @@ void dss_unregister_key(struct dss_module_key *key);
 
 #define DSS_XS_NAME_LEN		64
 
+struct srv_profile_chunk {
+	d_list_t	spc_chunk_list;
+	uint32_t	spc_chunk_offset;
+	uint32_t	spc_chunk_size;
+	uint64_t	*spc_chunks;
+};
+
+/* The profile structure to record single operation */
+struct srv_profile_op {
+	int		pro_op;			/* operation */
+	char		*pro_op_name;		/* name of the op */
+	int		pro_acc_cnt;		/* total number of val */
+	int		pro_acc_val;		/* current total val */
+	d_list_t	pro_chunk_list;		/* list of all chunks */
+	d_list_t	pro_chunk_idle_list;	/* idle list of profile chunk */
+	int		pro_chunk_total_cnt;	/* Count in idle list & list */
+	int		pro_chunk_cnt;		/* count in list */
+	struct srv_profile_chunk *pro_current_chunk; /* current chunk */
+};
+
+/* Holding the total trunk list for a specific profiling module */
+struct srv_profile {
+	struct srv_profile_op *sp_ops;
+	int		sp_ops_cnt;
+	int		sp_avg;
+	int		sp_id;
+	char		*sp_dir_path;	/* Where to dump the profiling */
+	char		**sp_names;	/* profile name */
+	ABT_thread	sp_dump_thread;	/* dump thread for profile */
```

Comment:
Suggestion: The dump thread looks not necessary to me, dump can happen on the profiling end. By removing this dump ULT, we can:
- Make the profile being able to be applied to standalone VOS and client stack;
- Avoid the race of profiling xstream vs. dumping xstream;
- Simplify code;

I don't think profiling will use too much memory and causing oom, if you worried the memory usage, just set a reasonable upper limit should be fine. (In your current implementation, the upper limit is 1MB).

Reply:
Well, the key is to be able to access the profile buffer in each function, standalone VOS might be ok. but not sure how to access the buffer on client stack easily.